x/contracts/examples/multisig/src/lib.rs

```diff
@@ -0,0 +1,480 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+use std::{collections::HashSet, num::NonZeroU32};
+
+use wasmlanche::{
+    borsh::{BorshDeserialize, BorshSerialize},
+    public, state_schema, Address, Context,
+};
+
+pub type ProposalId = usize;
+
+state_schema! {
+    ProposalCount => ProposalId,
+    ProposalItem(ProposalId) => Proposal,
+    ProposalVoteInfo(ProposalId) => ProposalMeta,
+    Vote(ProposalId, Address) => bool,
+}
+
+#[public]
+pub fn propose(
+    ctx: &mut Context,
+    proposal: Proposal,
+    voters: Voters,
+    quorum: NonZeroU32,
+) -> ProposalId {
+    let actor = ctx.actor();
+    let mut quorum = quorum.get();
+
+    if voters.contains(actor) {
+        quorum -= 1;
+    }
+
+    let quorum =
+        NonZeroU32::new(quorum).expect("cannot create a proposal that immediately executes");
+
+    let meta = voters
+        .with_quorum(quorum)
+        .expect("quorum larger than eligible voters");
+
+    let id = ctx
+        .get(ProposalCount)
+        .expect("failed to deserialize")
+        .unwrap_or_default();
+
+    ctx.store((
+        (ProposalCount, id + 1),
+        (ProposalItem(id), proposal),
+        (ProposalVoteInfo(id), meta),
+        (Vote(id, actor), true),
+    ))
+    .expect("failed to store proposal");
+
+    id
+}
+
+#[public]
+pub fn vote(ctx: &mut Context, proposal_id: ProposalId, vote: bool) -> Option<Vec<u8>> {
+    let actor = ctx.actor();
+
+    let previous_vote = ctx
+        .get(Vote(proposal_id, actor))
+        .expect("failed to deserialize");
+
+    if previous_vote.is_some() {
+        return None;
+    }
+
+    ctx.store_by_key(Vote(proposal_id, actor), vote)
+        .expect("failed to serialize");
+
+    if !vote {
+        return None;
+    }
+
+    let ProposalMeta {
+        voters,
+        votes_remaining,
+    } = ctx
+        .get(ProposalVoteInfo(proposal_id))
+        .expect("failed to deserialize")
+        .expect("proposal not found");
+
+    if !voters.contains(actor) || votes_remaining == 0 {
+        return None;
+    }
+
+    let votes_remaining = votes_remaining - 1;
+
+    ctx.store_by_key(
+        ProposalVoteInfo(proposal_id),
+        ProposalMeta {
+            voters,
+            votes_remaining,
+        },
+    )
+    .expect("failed to store votes remaining");
+
+    match votes_remaining {
+        0 => {
+            let Proposal {
+                contract_address,
+                function_name,
+                args,
+            } = ctx
+                .get(ProposalItem(proposal_id))
+                .expect("failed to deserialize")
+                .expect("proposal not found");
+            // TODO: fix this
+            // let remaining_fuel = ctx.remaining_fuel();
+            let remaining_fuel = 1_000_000;
+
+            let result: DeferDeserialization = ctx
+                .call_contract(contract_address, &function_name, &args, remaining_fuel, 0)
```

Comment:
the gas would run out, but could be a good way to test re-entrancy nonetheless

Reply:
Haha, yeah, I guess it would drain the gas of the quorum voter. In practice, I don't think this is a big deal because you would have to hard-code the vote, which would be an immediate red flag (assuming that you would audit the code that you're voting for).